### PR TITLE
VACMS-18855 Add Find Forms download modal back to form detail page

### PR DESCRIPTION
**Pull request by randimays**
## Summary
In [this PR](https://github.com/department-of-veterans-affairs/vets-website/pull/28068) we inadvertently removed the download modal from form detail pages when removing the feature toggle for the modal. This current PR plus [its companion content-build PR](https://github.com/department-of-veterans-affairs/content-build/pull/2239) add that modal back.

Includes a bit of cleanup/refactor covered in comments in the diffs.

If you want to run everything locally (including content-build) to go through the test cases, point your content-build `.env` to this DRUPAL_ADDRESS: https://cms-ciyktotgoqohism0kiuuviu13oifigrz.demo.cms.va.gov/. This will pull the correct feature flag in the "on" state.

### Gotchas

1. We lost some unit and end-to-end test coverage due to changing DOM elements.
2. There are local testing hurdles due to the way the API/HTTP calls work. They are described in the Loom walkthroughs in `Testing done`.

## Related issue(s)
- https://github.com/department-of-veterans-affairs/va.gov-cms/issues/18855

### Test case 1: Opening the modal on the search results page (feature toggle on)
✅ [Loom walkthrough](https://www.loom.com/share/f4145a2126af4aeca93af665b707d1d6)
<details><summary>Screenshots and test steps</summary>
<br/>

- Go to `/find-forms`
- Enter search term (I used `10-0003k`)
- Validate that "Download VA Form ____" under a search result is a `<button>`
- Click download button for one or more search results
- Validate that focus is on modal close button when modal opens
- Validate that modal shows up with two links:
  - Adobe link goes [here](https://get.adobe.com/reader/)
  - `Download VA Form ____` links to the **production** form (verify this same result in production to make sure link is correct)
- Click close button on modal and validate it closes and returns focus to the original modal trigger
- Repeat steps on mobile
<table>
<tr><th>Modal triggers in DOM (desktop and mobile)</th></tr>
<tr><td>Screenshot 2024-09-05 at 9 38 15 AM</td></tr>
</table>

### Desktop

<table>
<tr><th>Search results</th></tr>
<tr><td>Screenshot 2024-09-05 at 9 38 03 AM</td></tr>
</table>

<table>
<tr><th>Modal links in DOM</th><th>Modal visual</th></tr>
<tr><td>Screenshot 2024-09-05 at 9 38 53 AM</td><td>Screenshot 2024-09-05 at 9 38 24 AM</td></tr>
</table>

<table>
<tr><th>PDF that opens when you click the Download button in the modal</th></tr>
<tr><td>Screenshot 2024-09-05 at 9 39 13 AM</td></tr>
</table>

<hr />

### Mobile

<table>
<tr><th>Search results</th></tr>
<tr><td>Screenshot 2024-09-05 at 9 46 23 AM</td></tr>
</table>

<table>
<tr><th>Modal links in DOM</th><th>Modal visual</th></tr>
<tr><td>Screenshot 2024-09-05 at 9 46 42 AM</td><td>Screenshot 2024-09-05 at 9 49 18 AM</td></tr>
</table>

</details>

### Test case 2: Opening the modal on the form detail page with only one download button (feature toggle on)
✅ [Loom walkthrough](https://www.loom.com/share/2662fb7d07224c4a995045fcd844d0c8)
✅ [More technical (code) walkthrough](https://www.loom.com/share/7a515547d2694e29ad3f06b758c52d73) - This is how I got the modal to pop up locally from the form detail page. You won't be able to see the modal if you simply pull down this branch and run it
<details><summary>Screenshots and test steps</summary>
<br/>

- Go to `/find-forms`
- Enter search term (I used `10-10m`)
- Click on a search result to get to a form detail page (e.g. `/find-forms/about-form-10-10m/`)
- On form detail page under `Downloadable PDF`:
  - validate `Download VA Form ___ (PDF)` element is a `<button>`
  - click download button
- Expect a slight delay for the modal to pop up (this is cross comms between content-build and vets-website)
- Validate that focus is on modal close button when modal opens
- Validate that modal shows up with two links:
  - Adobe link goes [here](https://get.adobe.com/reader/)
  - Download link links to the **production** form (verify this same result in production to make sure link is correct)
- Click close button on modal and validate it closes and returns focus to the original modal trigger
- Repeat steps on mobile

<table>
<tr><th>Modal trigger in the DOM (desktop and mobile)</th></tr>
<tr><td>Screenshot 2024-08-14 at 2 07 52 PM</td></tr>
</table>

### Desktop

<table>
<tr><th>Form detail page</th></tr>
<tr><td>Screenshot 2024-09-05 at 12 25 04 PM</td></tr>
</table>

<table>
<tr><th>Modal links in DOM</th><th>Modal visual</th></tr>
<tr><td>Screenshot 2024-08-14 at 2 02 50 PM</td><td>Screenshot 2024-08-14 at 2 03 01 PM</td></tr>
</table>

<hr />

### Mobile 

<table>
<tr><th>Form detail page</th></tr>
<tr><td>Screenshot 2024-09-05 at 12 26 02 PM</td></tr>
</table>

<table>
<tr><th>Modal links in DOM</th><th>Modal visual</th></tr>
<tr><td>Screenshot 2024-08-14 at 2 04 24 PM</td><td>Screenshot 2024-08-14 at 2 04 12 PM</td></tr>
</table>
</details>

### Test case 3: Opening the modal on the form detail page with multiple download buttons (feature toggle on)

✅ [Loom walkthrough](https://www.loom.com/share/9b250b4238a345708a44e18bd501a9cb)

<details><summary>Screenshots and test steps</summary>
<br/>

- Go to `/find-forms`
- Enter search term for a form with related forms (I used 10-0003k here: `/find-forms/about-form-10-0003k/`)
- Click on the form title (the h3)
- On the form detail page under `Downloadable PDF`:
  - validate the `Download VA Form ___ (PDF)` element is a `<button>`
  - click the download button
  - Validate that modal shows up with two links:
    - Adobe link goes [here](https://get.adobe.com/reader/)
    - Download link links to the **environment-specific** form (same in prod; verify this same link in production to make sure link is correct).
    - Click close button on the modal and validate it closes and returns focus (yellow outline) to the button you originally clicked
- On the form detail page under `Related forms and instructions`:
  - Repeat the validation steps from the "Downloadable PDF" section
- Repeat the above steps on mobile

<table>
<tr><th>Modal triggers in DOM for 10-0003K (desktop and mobile)</th><th>Modal triggers in DOM for 10-0003K-2 </th></tr>
<tr><td>Screenshot 2024-09-05 at 12 17 12 PM</td><td>Screenshot 2024-09-05 at 12 17 24 PM</td></tr>
</table>

### Desktop

<table>
<tr><th>Form detail page</th></tr>
<tr><td>Screenshot 2024-09-05 at 12 29 08 PM</td></tr>
</table>

<table>
<tr><th>Modal links in DOM (10-0003K)</th><th>Modal visual (10-0003K)</th></tr>
<tr><td>Screenshot 2024-09-05 at 12 28 38 PM</td><td>Screenshot 2024-09-05 at 12 28 28 PM</td></tr>
</table>

<table>
<tr><th>Modal links in DOM (10-0003K-2)</th><th>Modal visual (10-0003K-2)</th></tr>
<tr><td>Screenshot 2024-09-05 at 12 33 28 PM</td><td>Screenshot 2024-09-05 at 12 33 21 PM</td></tr>
</table>

<hr/>

### Mobile

<table>
<tr><th>Form detail page</th></tr>
<tr><td>Screenshot 2024-09-05 at 12 30 41 PM</td></tr>
</table>

<table>
<tr><th>Modal links in DOM (10-0003K)</th><th>Modal visual (10-0003K)</th></tr>
<tr><td>Screenshot 2024-09-05 at 12 31 01 PM</td><td>Screenshot 2024-09-05 at 12 30 52 PM</td></tr>
</table>

<table>
<tr><th>Modal links in DOM (10-0003K-2)</th><th>Modal visual (10-0003K-2)</th></tr>
<tr><td>Screenshot 2024-09-05 at 12 31 08 PM</td><td>Screenshot 2024-09-05 at 12 31 15 PM</td></tr>
</table>

</details>

### Test case 4: Some kind of form error (feature toggle on)

✅ [Loom walkthrough](https://www.loom.com/share/0b7dc78b768243bd9435357a586a2e93)
<details><summary>Screenshots and test steps</summary>
<br/>

The form detail page is built largely by content-build. CB fetches the data from Drupal and the form URLs are formatted somewhere in the metalsmith / content-build process to be environment-specific. Because of that, the form URLs don't work when running the app locally. If you follow these steps:

- Go to `/find-forms`
- Enter search term (I used `10-10m`)
- Click a search result to get to a form detail page (e.g. `/find-forms/about-form-10-10m/`)
- On the form detail page under `Downloadable PDF`:
  - validate the `Download VA Form ___ (PDF)` element is a `<button>`
  - click the download button

<table>
<tr><th>Error message</th><th>DOM</th></tr>
<tr><td>Screenshot 2024-08-14 at 2 51 13 PM</td><td>Screenshot 2024-08-14 at 3 16 41 PM</td></tr>
</table>

</details>

### Test case 5: Some kind of form error (multiple download links) (feature toggle on)

✅ [Loom walkthrough](https://www.loom.com/share/c8aa0d565a1a44c78261bf14a0434f98)
<details><summary>Screenshots and test steps</summary>
<br/>

- Go to `/find-forms`
- Enter search term for a form with related forms (I used 10-0003k here: `/find-forms/about-form-10-0003k/`)
- On the form detail page under `Downloadable PDF`:
  - validate the `Download VA Form ___ (PDF)` element is a `<button>`
  - click the download button
- On the form detail page under `Related forms and instructions`:
  - validate the `Download VA Form ___ (PDF)` element is a `<button>`
  - click the download button

<table>
<tr><th>Error alerts</th></tr>
<tr><td>Screenshot 2024-09-05 at 12 56 09 PM</td></tr>
</table>

<table>
<tr><th>DOM for the Downloadable PDF error</th></tr>
<tr><td>Screenshot 2024-09-10 at 3 08 28 PM
</td></tr>
</table>

<table>
<tr><th>DOM for the Related form error</th></tr>
<tr><td>Screenshot 2024-09-10 at 3 08 37 PM</td></tr>
</table>

<table>
<tr><th>Mobile view</th></tr>
<tr><td>Screenshot 2024-09-05 at 12 56 09 PM</td></tr>
</table>

</details>

### Test case 6: Feature toggle is off

No Loom walkthrough.

When the feature toggle in content-build is off, the version of the form detail page that is currently in prod should show. This means that the download link under `Downloadable PDF` and under `Related forms and instructions` should be an `a` tag and the markup should match what is in prod today.

There is one exception: the `data-widget-type` does not appear on this newer version of the link because clicking a download **link** (`a` tag) does not also reliably trigger an `onClick` event, meaning the code in vets-website will never be triggered, and the `data-widget-type` is not used anyway.

Screenshot 2024-09-10 at 2 19 35 PM
Screenshot 2024-09-10 at 2 20 00 PM

## What areas of the site does it impact?

Find Forms search results and form detail pages.

### Quality Assurance & Testing

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] Linting warnings have been addressed
- [x] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [x] Browser console contains no warnings or errors.
- [x] Events are being sent to the appropriate logging solution

### CODE REVIEW

File: src/applications/find-forms/components/PdfAlert.jsx [hunk 1]
```diff
@@ -27,7 +27,7 @@ const PdfAlert = () => {
   }
 
   return ReactDOM.createPortal(
-    <va-alert uswds>
```
Comment (randimays) [inline src/applications/find-forms/components/PdfAlert.jsx:30]: `uswds` is the default now; I removed it from all the files it was present

File: src/applications/find-forms/components/PdfModal.jsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import React from 'react';
```
Comment (randimays) [inline src/applications/find-forms/components/PdfModal.jsx:1]: We had two separate modals in the code: one for the form search results page and the other for the form detail page. I consolidated those two into this file.

File: src/applications/find-forms/tests/widgets/DownloadPDFGuidance.unit.spec.js [hunk 3]
```diff
@@ -1,108 +0,0 @@
-import React from 'react';
```
Comment (randimays) [inline src/applications/find-forms/tests/widgets/DownloadPDFGuidance.unit.spec.js:1]: This file was renamed to `DownloadHandler.unit.spec.js`

File: src/applications/find-forms/tests/widgets/DownloadPDFModal.unit.spec.jsx [hunk 4]
```diff
@@ -1,34 +0,0 @@
-import React from 'react';
```
Comment (randimays) [inline src/applications/find-forms/tests/widgets/DownloadPDFModal.unit.spec.jsx:1]: This file was renamed to `PdfModal.unit.spec.jsx`

File: src/applications/find-forms/widgets/createFindVaFormsPDFDownloadHelper/DownloadHandler.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
+import React from 'react';
```
Comment (randimays) [inline src/applications/find-forms/widgets/createFindVaFormsPDFDownloadHelper/DownloadHandler.js:1]: This code is largely the same as what was here before (in `DownloadPDFGuidance`). I just renamed it.

File: src/applications/find-forms/components/PdfModal.jsx (hunk 2 above)
Comment (eselkin) [reply src/applications/find-forms/components/PdfModal.jsx:1]: Makes sense

**Review by eselkin — DISMISSED** [review]
Good consolidation. LGTM

**Review by chriskim2311 — DISMISSED** [review]
LGTM!

**Review by eselkin — DISMISSED** [review]
LGTM

**Review by chriskim2311 — DISMISSED** [review]
LGTM, went through the looms great work!

**Comment by laflannery** [comment]
@randimays I have 2 comments:
1. On the listing button can you remove the `tabindex`. I checked and it's there now so I just must have missed this but we don't need it. I also didn't see it on any of your other screenshot but if it's anywhere else it's not needed anywhere. Buttons are by default in the tab order and having this is just asking for trouble lol
2. This is a question because I'm not sure how difficult it is and I think it's slightly out of scope. Is it possible to change the heading level on the  alert under "Related forms and instructions" to be an h4 but keep the heading level on the other alert an h3? Currently the way it is now, it's not properly nested under the related form title.
image

**Comment by randimays** [comment]
@laflannery `tabindex` is removed and I adjusted the header hierarchy. Here are the screenshots for the alerts section with updated headers.

Screenshot 2024-09-11 at 4 52 17 PM

DOM for the "Downloadable PDF" section
Screenshot 2024-09-11 at 4 52 26 PM

DOM for the "Related forms" section
Screenshot 2024-09-11 at 4 55 31 PM

**Comment by randimays** [comment]
@eselkin @chriskim2311 I had to push a commit for PR feedback ([here's the commit](https://github.com/department-of-veterans-affairs/vets-website/pull/31395/commits/4253f79229addb3318a7c9e2a2f2adb2bcf33b88); these are the only changes) - mind re-reviewing when you have a chance please?

**Review by chriskim2311 — APPROVED** [review]
LGTM!

**Review by powellkerry — APPROVED** [review]

**Comment by laflannery** [comment]
Screenshots look good, approved!

**Pull request merged**